client/components/Editor/plugins/equationsToMath.ts

```diff
@@ -0,0 +1,58 @@
+import { Node } from 'prosemirror-model';
+import { EditorState, Plugin } from 'prosemirror-state';
+import { EditorView } from 'prosemirror-view';
+
+type Replacement = {
+	pos: number;
+	replacementNode: Node;
+};
+
+const getEquationTranslationTransactionForState = (editorState: EditorState) => {
+	const transaction = editorState.tr;
+	const replacements: Replacement[] = [];
+	let mustReturnTransaction = false;
+	const mathInlineType = editorState.schema.nodes.math_inline;
+	const mathDisplayType = editorState.schema.nodes.math_display;
+	editorState.doc.descendants((node, pos) => {
+		const replacementNodeType =
+			node.type.name === 'equation'
+				? mathInlineType
+				: node.type.name === 'block_equation'
+				? mathDisplayType
+				: null;
+		if (replacementNodeType) {
+			const replacementNode = replacementNodeType.create(
+				null,
+				editorState.schema.text(node.attrs.value),
+			);
+			replacements.push({ pos, replacementNode });
+			mustReturnTransaction = true;
+		}
+	});
+	if (mustReturnTransaction) {
+		replacements
+			.reverse()
+			.forEach(({ pos, replacementNode }) =>
+				transaction.replaceWith(pos, pos + 1, replacementNode),
+			);
+		return transaction;
+	}
+	return null;
+};
+
+export default (_, props) => {
+	if (props.isReadOnly) {
+		return [];
+	}
+	return new Plugin({
```

Comment:
To reduce the overhead of this plugin iterating through all the nodes on every transaction, it might be worth trying to make sure it only runs once (it shouldn't need to again!). A sketch of how this would work follows.

The first step is to instantiate a "plugin key". This is basically a unique, never-colliding identifier for this plugin to use when it stores and retrieves information from shared PM state (in `EditorState` and `Transaction` objects):

```ts
const equationsToMathPluginKey = new PluginKey("equations-to-math");
```

Then modify the `Plugin` constructor to hold a state value called `hasRun`. This starts `false` and is updated every transaction — it becomes `true` if the `equationsToMathPluginKey` has added anything to the [transaction metadata](https://prosemirror.net/docs/guide/#state.plugins):

```ts
new Plugin({
    view: ... // as-is,
    appendTransaction: ... // as-is,
    key: equationsToMathPluginKey,
    state: {
        init: () => {
            return { hasRun: false }
        },
        apply: (transaction: Transaction, { hasRun }) => {
            return { hasRun: transaction.getMeta(equationsToMathPluginKey) || hasRun };
        },
    }
})
```

Finally, modify `getEquationTranslationTransactionForState` to (1) signal that it has run and (2) short-circuit if it turns out that it's already run:

```ts
const getEquationTranslationTransactionForState = (editorState: EditorState) => {
    // Return early if the plugin state says we already did the translation.
    const pluginState = equationsToMathPluginKey.getState(editorState);
    if (pluginState.hasRun) {
         return null;
    }
    // ... some time later ...
    if (mustReturnTransaction) {
        // Signal back to the plugin state that this transaction handled the node translations
        transaction.setMeta(equationsToMathPluginKey, true);
    }
};
```